src/engine/client/audio.cpp

```diff
@@ -328,3 +317,52 @@ void AudioManager::PlayGlobalSound(const std::string &path) {
     global_source_.reset(CreateSource(path));
     global_source_->StartPlaying();
 }
+
+void AudioManager::SilenceSources() {
+    for(auto sound : sources_) {
+        sound->StopPlaying();
+    }
+
+    global_source_->StopPlaying();
+}
+
+void AudioManager::FreeSources() {
+    LogInfo("freeing all sources...\n");
+
+    SilenceSources();
+    sources_.clear();
+}
+
+void AudioManager::FreeSamples(bool force) {
+    LogInfo("freeing all samples...\n");
+    if(force) {
+        /* clears absolutely everything */
+        samples_.clear();
+    } else {
+        /* clears only those not marked with preserve */
+        for(auto sample : samples_) {
```

Comment:
```
for(auto sample = samples_.begin(); sample != samples_.end();)
{
    if(!sample.second.preserve_) {
        sample = samples_.erase(sample);
    } else {
        ++sample;
    }
}
```

Erasing from a map invalidates any iterators to that element, so it'll be undefined behaviour when the loop tries to reach the next element from the now-invalid iterator.